sqlite3.go

```diff
@@ -1782,26 +1782,20 @@ func (d *SQLiteDriver) Open(dsn string) (driver.Conn, error) {
 }
 
 // Close the connection.
-func (c *SQLiteConn) Close() error {
+func (c *SQLiteConn) Close() (err error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if c.db == nil {
+		return nil // Already closed
+	}
+	runtime.SetFinalizer(c, nil)
 	rv := C.sqlite3_close_v2(c.db)
 	if rv != C.SQLITE_OK {
-		return c.lastError()
+		err = c.lastError()
```

Comment:
Given that the only way for `sqlite3_close_v2` to fail is for it to be called more than once (in which case it returns `SQLITE_MISUSE`), I'm not sure that calling `c.lastError()` here is safe in general, given that `sqlite3_errcode` and  `sqlite3_extended_errcode` probably have undefined behavior at that point. It's probably safer to instead directly create an `Error` that contains only the `Code` (with explanatory comments).

Reply:
That's `sqlite3_close`, not `sqlite3_close_v2`.

Reply:
> Calls to sqlite3_close() and sqlite3_close_v2() return [SQLITE_OK](https://www.sqlite.org/rescode.html#ok) if the [sqlite3](https://www.sqlite.org/c3ref/sqlite3.html) object is successfully destroyed and all associated resources are deallocated.

Yep (which makes sense given the context of the CLI - single-threaded and no GC), but going by the docs any error here means that the database was not fully closed and thus we should be able to safely call `sqlite3_errcode` and friends with it.

> If sqlite3_close_v2() is called with unfinalized prepared statements, unclosed BLOB handlers, and/or unfinished sqlite3_backups, it returns SQLITE_OK regardless, but instead of deallocating the database connection immediately, it marks the database connection as an unusable "zombie" and makes arrangements to automatically deallocate the database connection after all prepared statements are finalized, all BLOB handles are closed, and all backups have finished. The sqlite3_close_v2() interface is intended for use with host languages that are garbage collected, and where the order in which destructors are called is arbitrary.

The real question here and one that I don't know the answer to is: under what circumstances will close fail and set a useful extended error code on the db conn object? 

If the answer to that question is **never** than we can omit the check for the extended error code, but taking a look at the docs it seems a bit ambiguous since they state the conditions under which `sqlite3_close_v2` will return SQLITE_OK but not the conditions that will lead to other error codes.

That said, most (if not all code) just wants to know if the close succeeded or failed (then report the error without asserting on the failure cause) so I'm happy to make the change if it gets this PR over the line and revisit this if any issues or further discovery show that we actually should be asking for the extended error code.